File: backend/app/services/admin_service.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta
import time
from typing import Any

from sqlalchemy import func
from sqlalchemy.orm import Session

from app.models import (
    ActivityLog,
    Advisory,
    Alert,
    AnalyticsEvent,
    Crop,
    Equipment,
    EquipmentBooking,
    FarmImage,
    Farmer,
    FarmingTechnique,
    User,
)

APP_START_TIME = time.time()


@dataclass
class AdminService:
    """Service layer for admin analytics and management APIs."""

    def get_dashboard_stats(self, db: Session) -> dict[str, Any]:
        now = datetime.utcnow()
        seven_days_ago = now - timedelta(days=7)
        day_start = datetime(now.year, now.month, now.day)

        total_users = db.query(func.count(User.id)).scalar() or 0
        active_users = db.query(func.count(User.id)).filter(User.last_active >= seven_days_ago).scalar() or 0
        total_farmers = db.query(func.count(User.id)).filter(User.role == 'farmer').scalar() or 0
        active_crops = db.query(func.count(Crop.id)).scalar() or 0
        dau = db.query(func.count(User.id)).filter(User.last_active >= day_start).scalar() or 0

        growth_rows = (
            db.query(func.date(User.last_active).label('date'), func.count(User.id).label('count'))
            .filter(User.last_active.isnot(None), User.last_active >= seven_days_ago)
            .group_by(func.date(User.last_active))
            .order_by(func.date(User.last_active))
            .all()
        )
        growth_map = {str(row.date): int(row.count) for row in growth_rows if row.date is not None}

        user_growth_data: list[dict[str, Any]] = []
        for day_offset in range(6, -1, -1):
            day = now - timedelta(days=day_offset)
            key = day.date().isoformat()
            user_growth_data.append({'date': key, 'count': growth_map.get(key, 0)})

        crop_rows = (
            db.query(Crop.name.label('name'), func.count(Crop.id).label('count'))
            .group_by(Crop.name)
            .order_by(func.count(Crop.id).desc())
            .limit(10)
            .all()
        )

        feature_usage_counts = {
            'alertsCount': db.query(func.count(Alert.id)).scalar() or 0,
            'equipmentBookingsCount': db.query(func.count(EquipmentBooking.id)).scalar() or 0,
            'techniquesCount': db.query(func.count(FarmingTechnique.id)).scalar() or 0,
        }
        feature_usage_chart = [
            {'feature': 'alerts', 'count': int(feature_usage_counts['alertsCount'])},
            {'feature': 'equipmentBookings', 'count': int(feature_usage_counts['equipmentBookingsCount'])},
            {'feature': 'techniques', 'count': int(feature_usage_counts['techniquesCount'])},
        ]

        crop_distribution = [
            {'cropType': row.name, 'name': row.name, 'count': int(row.count)} for row in crop_rows
        ]

        return {
            'totalUsers': int(total_users),
            'activeUsers': int(active_users),
            'totalFarmers': int(total_farmers),
            'activeCrops': int(active_crops),
            'dau': int(dau),
            'userGrowthData': user_growth_data,
            'cropDistribution': crop_distribution,
            'featureUsage': feature_usage_chart,
            'featureUsageCounts': feature_usage_counts,
        }

    def get_farming_insights(self, db: Session) -> dict[str, Any]:
        top_crops = (
            db.query(Crop.name.label('name'), func.count(Crop.id).label('count'))
            .group_by(Crop.name)
            .order_by(func.count(Crop.id).desc())
            .limit(8)
            .all()
        )

        techniques = (
            db.query(FarmingTechnique.title.label('name'), func.count(FarmingTechnique.id).label('count'))
            .group_by(FarmingTechnique.title)
            .order_by(func.count(FarmingTechnique.id).desc())
            .limit(8)
            .all()
        )

        region_wise = (
            db.query(User.region.label('region'), func.count(User.id).label('count'))
            .filter(User.role == 'farmer')
            .group_by(User.region)
            .order_by(func.count(User.id).desc())
            .all()
        )

        return {
            'topCrops': [{'name': row.name, 'count': int(row.count)} for row in top_crops],
            'mostUsedTechniques': [{'name': row.name, 'count': int(row.count)} for row in techniques],
            'regionWiseFarmers': [
                {'region': row.region or 'Unknown', 'count': int(row.count)} for row in region_wise
            ],
        }

    def get_analytics(self, db: Session) -> dict[str, Any]:
        now = datetime.utcnow()
        day_start = now - timedelta(days=1)
        month_start = now - timedelta(days=30)

        dau = db.query(func.count(User.id)).filter(User.last_active >= day_start).scalar() or 0
        mau = db.query(func.count(User.id)).filter(User.last_active >= month_start).scalar() or 0

        prev_window_end = month_start
        prev_window_start = prev_window_end - timedelta(days=30)
        retained_users = (
            db.query(func.count(User.id))
            .filter(User.last_active >= month_start, User.created_at <= prev_window_end)
            .scalar()
            or 0
        )
        baseline_users = (
            db.query(func.count(User.id))
            .filter(User.created_at >= prev_window_start, User.created_at < prev_window_end)
            .scalar()
            or 0
        )
        retention_rate = round((retained_users / baseline_users) * 100, 2) if baseline_users else 0.0

        feature_events = (
            db.query(AnalyticsEvent.feature.label('feature'), func.count(AnalyticsEvent.id).label('count'))
            .filter(AnalyticsEvent.created_at >= month_start)
            .group_by(AnalyticsEvent.feature)
            .order_by(func.count(AnalyticsEvent.id).desc())
            .all()
        )

        if not feature_events:
            feature_events = [
                ('alerts', db.query(func.count(Alert.id)).scalar() or 0),
                ('equipment', db.query(func.count(EquipmentBooking.id)).scalar() or 0),
                ('techniques', db.query(func.count(FarmingTechnique.id)).scalar() or 0),
            ]
            usage = [{'feature': row[0], 'count': int(row[1])} for row in feature_events]
        else:
            usage = [{'feature': row.feature or 'unknown', 'count': int(row.count)} for row in feature_events]

        return {
            'dau': dau,
            'mau': mau,
            'retentionRate': retention_rate,
            'retentionData': {
                'retainedUsers': retained_users,
                'baselineUsers': baseline_users,
                'rate': retention_rate,
            },
            'featureUsage': usage,
        }

    def get_equipment_analytics(self, db: Session) -> dict[str, Any]:
        total = db.query(func.count(Equipment.id)).scalar() or 0
        total_bookings = db.query(func.count(EquipmentBooking.id)).scalar() or 0
        booked = (
            db.query(func.count(EquipmentBooking.id))
            .filter(EquipmentBooking.status.in_(['confirmed', 'approved', 'pending']))
            .scalar()
            or 0
        )

        most_rented_rows = (
            db.query(Equipment.name.label('name'), func.count(EquipmentBooking.id).label('count'))
            .join(EquipmentBooking, Equipment.id == EquipmentBooking.equipment_id)
            .group_by(Equipment.name)
            .order_by(func.count(EquipmentBooking.id).desc())
            .limit(10)
            .all()
        )

        return {
            'totalListings': total,
            'totalBookings': total_bookings,
            'activeBookings': booked,
            'mostRentedEquipment': [{'name': row.name, 'count': int(row.count)} for row in most_rented_rows],
        }

    def get_system_health(self, db: Session) -> dict[str, Any]:
        now = datetime.utcnow()
        one_day_ago = now - timedelta(days=1)

        recent_errors = (
            db.query(func.count(ActivityLog.id))
            .filter(ActivityLog.level == 'error', ActivityLog.timestamp >= one_day_ago)
            .scalar()
            or 0
        )
        uptime_seconds = max(0, int(time.time() - APP_START_TIME))

        return {
            'serverTime': now.isoformat(),
            'uptime': uptime_seconds,
            'status': 'OK',
            'totalErrors': int(recent_errors),
            # Keep legacy fields for backward compatibility with existing screens.
            'apiStatus': 'ok',
            'databaseStatus': 'ok',
            'recentErrorCount': int(recent_errors),
        }


admin_service = AdminService()

File: src/app/core/services/admin.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';

import { environment } from '../../../environments/environment';

// Admin Models
export interface AdminUser {
  id: number;
  full_name: string;
  email: string;
  username: string;
  role: string;
  status?: string;
  region?: string | null;
  last_active?: string | null;
}

export interface DashboardStats {
  totalUsers: number;
  activeUsers: number;
  totalFarmers: number;
  activeCrops: number;
  dau: number;
  userGrowthData: Array<{ date: string; count: number }>;
  cropDistribution: Array<{ cropType?: string; name: string; count: number }>;
  featureUsage: Array<{ feature: string; count: number }>;
  featureUsageCounts?: {
    alertsCount: number;
    equipmentBookingsCount: number;
    techniquesCount: number;
  };
}

export interface FarmingInsights {
  topCrops: Array<{ name: string; count: number }>;
  mostUsedTechniques: Array<{ name: string; count: number }>;
  regionWiseFarmers: Array<{ region: string; count: number }>;
}

export interface AdvancedAnalytics {
  dau: number;
  mau: number;
  retentionRate: number;
  featureUsage: Array<{ feature: string; count: number }>;
}

export interface AdminAlert {
  id: number;
  expert_id: number;
  title: string;
  description: string;
  message?: string | null;
  status: string;
  alert_type: string;
  severity: string;
  target_regions: string;
  affected_crops?: string | null;
  recommendations?: string | null;
  created_at: string;
}

export interface EquipmentAnalytics {
  totalListings: number;
  totalBookings: number;
  activeBookings: number;
  mostRentedEquipment: Array<{ name: string; count: number }>;
}

export interface FarmImageRecord {
  id: number;
  user_id: number;
  image_url: string;
  crop_type?: string | null;
  uploaded_at: string;
}

export interface SystemHealth {
  status: string;
  serverTime: string;
  uptime: number;
  totalErrors: number;
  apiStatus?: string;
  databaseStatus?: string;
  recentErrorCount?: number;
}

export interface ExpertAssignment {
  id: number;
  advisory_id: number;
  expert_id: number;
  assigned_by: number;
  status: string;
  assignment_date: string;
  completion_date: string | null;
}

export interface Content {
  id: number;
  title: string;
  description: string;
  category: string;
  location: string | null;
  metadata: string | null;
  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface Equipment {
  id: number;
  name: string;
  description: string;
  daily_rent: number;
  location: string;
  owner_id: number;
  is_available: boolean;
  condition: string;
  created_at: string;
}

export interface ActivityLog {
  id: number;
  user_id: number;
  action: string;
  entity_type: string;
  entity_id: number | null;
  description: string;
  ip_address: string | null;
  timestamp: string;
}

export interface SystemAnalytics {
  total_users: number;
  total_farmers: number;
  total_experts: number;
  total_advisories: number;
  pending_advisories: number;
  avg_expert_response_time_hours: number;
  avg_farmer_satisfaction: number;
  active_advisories: number;
}

export interface AdminStats {
  total_farmers: number;
  total_experts: number;
  total_jobs: number;
  total_equipment: number;
  approved_equipment: number;
  pending_equipment: number;
}

export interface EmploymentJob {
  id: number;
  farmer_id: number;
  farmer_name: string;
  title: string;
  description: string;
  location: string;
  wage: number;
  duration?: string | null;
  status: 'open' | 'active' | 'closed' | string;
  created_at: string;
}

export interface EmploymentApplication {
  id: number;
  job_id: number;
  job_title: string;
  applicant_id: number;
  applicant_name: string;
  applicant_email: string;
  farmer_name: string;
  message: string | null;
  status: 'pending' | 'accepted' | 'rejected' | string;
  created_at: string;
}

export interface EmploymentWorker {
  id: number;
  name: string;
  contact: string;
  experience: string | null;
  is_blocked: boolean;
  created_at: string;
}

@Injectable({
  providedIn: 'root',
})
export class AdminService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/admin`;

  // ====================
  // USER MANAGEMENT
  // ====================

  getAllUsers(filters?: { role?: string; region?: string; status?: string }) {
    let params = new HttpParams();
    if (filters?.role) {
      params = params.set('role', filters.role);
    }
    if (filters?.region) {
      params = params.set('region', filters.region);
    }
    if (filters?.status) {
      params = params.set('status_filter', filters.status);
    }
    return this.http.get<AdminUser[]>(`${this.baseUrl}/users`, { params });
  }

  getUsersByRole(role: string) {
    return this.http.get<AdminUser[]>(`${this.baseUrl}/users/role/${role}`);
  }

  deleteUser(userId: number) {
    return this.http.delete(`${this.baseUrl}/users/${userId}`);
  }

  updateUserStatus(userId: number, status: 'active' | 'suspended') {
    return this.http.patch<{ id: number; status: string }>(`${this.baseUrl}/users/${userId}/status`, { status });
  }

  getUserActivity(userId: number, limit: number = 50) {
    const params = new HttpParams().set('limit', String(limit));
    return this.http.get<ActivityLog[]>(`${this.baseUrl}/users/${userId}/activity`, { params });
  }

  getDashboardStats() {
    return this.http.get<DashboardStats>(`${this.baseUrl}/dashboard-stats`);
  }

  getFarmingInsights() {
    return this.http.get<FarmingInsights>(`${this.baseUrl}/farming-insights`);
  }

  getAdvancedAnalytics() {
    return this.http.get<AdvancedAnalytics>(`${this.baseUrl}/analytics`);
  }

  createAlert(payload: {
    expert_id: number;
    title: string;
    description: string;
    message?: string;
    alert_type: string;
    severity?: string;
    target_regions: string;
    affected_crops?: string;
    recommendations?: string;
    expiry_date?: string;
  }) {
    return this.http.post<AdminAlert>(`${this.baseUrl}/alerts`, payload);
  }

  approveAlert(alertId: number) {
    return this.http.patch<AdminAlert>(`${this.baseUrl}/alerts/${alertId}/approve`, {});
  }

  getAlerts(status?: string) {
    let params = new HttpParams();
    if (status) {
      params = params.set('status_filter', status);
    }
    return this.http.get<AdminAlert[]>(`${this.baseUrl}/alerts`, { params });
  }

  getEquipmentAnalytics() {
    return this.http.get<EquipmentAnalytics>(`${this.baseUrl}/equipment/analytics`);
  }

  getFarmImages(userId?: number, limit: number = 100) {
    let params = new HttpParams().set('limit', String(limit));
    if (userId !== undefined) {
      params = params.set('user_id', String(userId));
    }
    return this.http.get<FarmImageRecord[]>(`${this.baseUrl}/farm-images`, { params });
  }

  getLogs(level?: string, days: number = 7, limit: number = 200) {
    let params = new HttpParams().set('days', String(days)).set('limit', String(limit));
    if (level) {
      params = params.set('level', level);
    }
    return this.http.get<ActivityLog[]>(`${this.baseUrl}/logs`, { params });
  }

  getSystemHealth() {
    return this.http.get<SystemHealth>(`${this.baseUrl}/system-health`);
  }

  // ====================
  // EXPERT ASSIGNMENT
  // ====================

  getExpertAssignments(status?: string) {
    let params = new HttpParams();
    if (status) {
      params = params.set('status_filter', status);
    }
    return this.http.get<ExpertAssignment[]>(`${this.baseUrl}/expert-assignments`, { params });
  }

  createExpertAssignment(advisoryId: number, expertId: number) {
    return this.http.post<ExpertAssignment>(`${this.baseUrl}/expert-assignments`, {
      advisory_id: advisoryId,
      expert_id: expertId,
    });
  }

  updateExpertAssignment(assignmentId: number, status: string) {
    return this.http.patch<ExpertAssignment>(
      `${this.baseUrl}/expert-assignments/${assignmentId}`,
      null,
      { params: new HttpParams().set('new_status', status) }
    );
  }

  // ====================
  // CONTENT MANAGEMENT
  // ====================

  getContents(category?: string) {
    let params = new HttpParams();
    if (category) {
      params = params.set('category', category);
    }
    return this.http.get<Content[]>(`${this.baseUrl}/contents`, { params });
  }

  createContent(data: {
    title: string;
    description: string;
    category: string;
    location?: string;
    metadata?: string;
  }) {
    return this.http.post<Content>(`${this.baseUrl}/contents`, data);
  }

  updateContent(
    contentId: number,
    data: {
      title: string;
      description: string;
      category: string;
      location?: string;
      metadata?: string;
    }
  ) {
    return this.http.put<Content>(`${this.baseUrl}/contents/${contentId}`, data);
  }

  deleteContent(contentId: number) {
    return this.http.delete(`${this.baseUrl}/contents/${contentId}`);
  }

  // ====================
  // EQUIPMENT MANAGEMENT
  // ====================

  getEquipment(isAvailable?: boolean) {
    let params = new HttpParams();
    if (isAvailable !== undefined) {
      params = params.set('is_available', isAvailable.toString());
    }
    return this.http.get<Equipment[]>(`${this.baseUrl}/equipment`, { params });
  }

  createEquipment(
    data: {
      name: string;
      description: string;
      daily_rent: number;
      location: string;
      condition?: string;
    },
    ownerId: number
  ) {
    const params = new HttpParams().set('owner_id', ownerId.toString());
    return this.http.post<Equipment>(`${this.baseUrl}/equipment`, data, { params });
  }

  // ====================
  // ACTIVITY MONITORING
  // ====================

  getActivities(userId?: number, action?: string, days: number = 7) {
    let params = new HttpParams().set('days', days.toString());
    if (userId) {
      params = params.set('user_id', userId.toString());
    }
    if (action) {
      params = params.set('action', action);
    }
    return this.http.get<ActivityLog[]>(`${this.baseUrl}/activities`, { params });
  }

  // ====================
  // ANALYTICS
  // ====================

  getSystemAnalytics() {
    return this.http.get<SystemAnalytics>(`${this.baseUrl}/analytics/system`);
  }

  getAdminStats() {
    return this.http.get<AdminStats>(`${this.baseUrl}/analytics/stats`);
  }

  // ====================
  // EMPLOYMENT / LABOR MANAGEMENT
  // ====================

  getEmploymentJobs(status?: string) {
    let params = new HttpParams();
    if (status) {
      params = params.set('status_filter', status);
    }
    return this.http.get<EmploymentJob[]>(`${this.baseUrl}/jobs`, { params });
  }

  getEmploymentApplications(status?: string) {
    let params = new HttpParams();
    if (status) {
      params = params.set('status_filter', status);
    }
    return this.http.get<EmploymentApplication[]>(`${this.baseUrl}/applications`, { params });
  }

  updateEmploymentApplicationStatus(applicationId: number, status: 'accepted' | 'rejected') {
    return this.http.put<EmploymentApplication>(`${this.baseUrl}/applications/${applicationId}/status`, { status });
  }

  getEmploymentWorkers(blocked?: boolean) {
    let params = new HttpParams();
    if (blocked !== undefined) {
      params = params.set('blocked', blocked.toString());
    }
    return this.http.get<EmploymentWorker[]>(`${this.baseUrl}/workers`, { params });
  }

  removeOrBlockWorker(workerId: number, hardDelete: boolean = false) {
    const params = new HttpParams().set('hard_delete', hardDelete.toString());
    return this.http.delete(`${this.baseUrl}/workers/${workerId}`, { params });
  }
}
